src/node/tendermint_processes.ml

```diff
@@ -0,0 +1,367 @@
+    { time; step; on_timeout; started = false; height; round }
+end
+
+type consensus_action =
+  | Broadcast         of sidechain_consensus_op
+  | Schedule          of consensus_action option Clock.t
+  | RestartTendermint of (height * round)
+  | DoNothing
+
+type process =
+  height ->
+  round ->
+  consensus_state ->
+  input_log ->
+  OutputLog.t ->
+  State.t ->
+  consensus_action option
+
+type clock = consensus_action option Clock.t
+
+let on_timeout_propose (height : height) (round : round)
+    (consensus_state : consensus_state) (msg_log : input_log) _dlog
+    _global_state =
+  let do_something consensus_state =
+    if
+      height = consensus_state.height
+      && round = consensus_state.round
+      && consensus_state.step = Proposal
+    then (
+      consensus_state.step <- Prevote;
+      Broadcast (PrevoteOP (consensus_state.height, consensus_state.round, nil)))
+    else
+      DoNothing in
+  if contains_timeout msg_log height Proposal then
+    Some (do_something consensus_state)
+  else
+    None
+
+let on_timeout_prevote (height : height) (round : round)
+    (consensus_state : consensus_state) (msg_log : input_log) _dlog
+    _global_state =
+  let do_something consensus_state =
+    if
+      height = consensus_state.height
+      && round = consensus_state.round
+      && consensus_state.step = Prevote
+    then (
+      consensus_state.step <- Precommit;
+      Broadcast
+        (PrecommitOP (consensus_state.height, consensus_state.round, nil)))
+    else
+      DoNothing in
+  if contains_timeout msg_log height Prevote then
+    Some (do_something consensus_state)
+  else
+    None
+
+let start_round (height : height) (round : round)
+    (consensus_state : consensus_state) (_msg_log : input_log) _dlog
+    global_state =
+  consensus_state.round <- round;
+  consensus_state.step <- Proposal;
+  (* If we start a round > 0 and already have a valid/locked value, we update their hash
+     with the correct round *)
+  (if consensus_state.valid_value <> nil then
+     let new_value = update_value consensus_state.valid_value round in
+     consensus_state.valid_value <- new_value);
+  (if consensus_state.locked_value <> nil then
+     let new_value = update_value consensus_state.locked_value round in
+     consensus_state.locked_value <- new_value);
+  let return_action =
+    if i_am_proposer global_state height round then
+      if consensus_state.valid_value <> nil then
+        Some
+          (Broadcast
+             (ProposalOP
+                ( consensus_state.height,
+                  consensus_state.round,
+                  consensus_state.valid_value,
+                  consensus_state.valid_round )))
+      else
+        Some
+          (Broadcast
+             (ProposalOP
+                ( consensus_state.height,
+                  consensus_state.round,
+                  (* FIXME: remove depdendency to full global_state *)
+                  !produce_value global_state,
```

Comment:
In case we're the producer for this round but there's no `valid_value` from last round (e.g., we're in a new round), we have to produce one. 

Sidenote: if you observe the algorithm, there shouldn't be a case where we have a valid_value that isn't nil, and where locked_value is. @AngryStitch this is an opportunity for a test! :p

Reply:
Yes. I'm gonna enforce stuff on the new consensus repo. Thanks :)